File: android_app/app/src/main/java/com/oceansentinels/app/data/remote/dto/AnalyticsDto.kt
package com.oceansentinels.app.data.remote.dto

import com.google.gson.annotations.SerializedName
import com.oceansentinels.app.domain.model.*

/**
 * Analytics DTOs for API communication
 * Aligned with actual backend response formats
 */

data class DashboardAnalyticsDto(
    @SerializedName("total_incidents") val totalIncidents: Int = 0,
    @SerializedName("active_incidents") val activeIncidents: Int = 0,
    @SerializedName("resolved_incidents") val resolvedIncidents: Int = 0,
    @SerializedName("false_alarms") val falseAlarms: Int = 0,
    @SerializedName("incidents_by_type") val incidentsByType: Map<String, Int>? = null,
    @SerializedName("recent_incidents") val recentIncidents: Int = 0,
    @SerializedName("average_response_time_hours") val averageResponseTimeHours: Double = 0.0,
    @SerializedName("last_updated") val lastUpdated: String? = null
) {
    fun toDomain(): DashboardAnalytics {
        // Derive pending = active - resolved (active includes PENDING, VERIFIED, IN_PROGRESS)
        val pendingCount = activeIncidents
        return DashboardAnalytics(
            totalIncidents = totalIncidents,
            activeIncidents = activeIncidents,
            resolvedIncidents = resolvedIncidents,
            totalRescueTeams = 0,
            totalAuthorities = 0,
            totalCitizens = 0,
            activeAlerts = activeIncidents,
            pendingCount = pendingCount,
            resolvedCount = resolvedIncidents,
            avgResponseTime = averageResponseTimeHours
        )
    }
}

data class TimelineDataPointDto(
    @SerializedName("date") val date: String? = null,
    @SerializedName("count") val count: Int = 0
) {
    fun toDomain(): TimelineDataPoint {
        return TimelineDataPoint(
            date = date ?: "",
            count = count
        )
    }
}

/**
 * Backend returns: {"timeline": [...], "period_days": int, "start_date": str, "end_date": str}
 */
data class IncidentsTimelineDto(
    @SerializedName("timeline") val timeline: List<TimelineDataPointDto>? = null,
    @SerializedName("period_days") val periodDays: Int = 30,
    @SerializedName("start_date") val startDate: String? = null,
    @SerializedName("end_date") val endDate: String? = null
) {
    fun toDomain(): IncidentsTimeline {
        val points = timeline?.map { it.toDomain() } ?: emptyList()
        return IncidentsTimeline(
            dataPoints = points,
            totalCount = points.sumOf { it.count },
            periodDays = periodDays
        )
    }
}

/**
 * Backend returns flat maps: {"by_type": {"HIGH_WAVES": 5}, "by_status": {...}, "by_urgency": {...}}
 */
data class IncidentsDistributionDto(
    @SerializedName("by_type") val byType: Map<String, Int>? = null,
    @SerializedName("by_status") val byStatus: Map<String, Int>? = null,
    @SerializedName("by_urgency") val byUrgency: Map<String, Int>? = null
) {
    fun toDomain(): IncidentsDistribution {
        return IncidentsDistribution(
            byStatus = mapToDistributionItems(byStatus),
            byHazardType = mapToDistributionItems(byType),
            byUrgency = mapToDistributionItems(byUrgency)
        )
    }

    private fun mapToDistributionItems(map: Map<String, Int>?): List<com.oceansentinels.app.domain.model.DistributionItem> {
        if (map == null) return emptyList()
        val total = map.values.sum().toFloat().coerceAtLeast(1f)
        return map.map { (label, value) ->
            com.oceansentinels.app.domain.model.DistributionItem(
                label = label.replace("_", " ").lowercase()
                    .replaceFirstChar { it.uppercase() },
                value = value,
                percentage = (value / total) * 100f
            )
        }
    }
}

/**
 * Backend returns: {"incidents": [{"lat", "lng", "type", "status"}], "total_located_incidents": int}
 */
data class GeographicIncidentDto(
    @SerializedName("lat") val lat: Double = 0.0,
    @SerializedName("lng") val lng: Double = 0.0,
    @SerializedName("type") val type: String? = null,
    @SerializedName("status") val status: String? = null
)

data class GeographicAnalyticsDto(
    @SerializedName("incidents") val incidents: List<GeographicIncidentDto>? = null,
    @SerializedName("total_located_incidents") val totalLocatedIncidents: Int = 0
) {
    fun toDomain(): GeographicAnalytics {
        val geoPoints = incidents?.map { incident ->
            GeographicData(
                regionName = incident.type?.replace("_", " ")?.lowercase()
                    ?.replaceFirstChar { it.uppercase() } ?: "Unknown",
                incidentCount = 1,
                latitude = incident.lat,
                longitude = incident.lng
            )
        } ?: emptyList()
        return GeographicAnalytics(
            regions = geoPoints,
            topRegions = geoPoints.take(5)
        )
    }
}

/**
 * Generic message response from backend (verify/deploy/resolve endpoints)
 */
data class MessageResponseDto(
    @SerializedName("message") val message: String = ""
)

File: IOS/OceanSentinels/Data/Remote/DTO/AnalyticsDTO.swift
import Foundation

// MARK: - Analytics DTOs

struct DashboardAnalyticsDTO: Codable {
    let totalIncidents: Int
    let activeIncidents: Int
    let resolvedIncidents: Int
    let falseAlarms: Int
    let incidentsByType: [String: Int]?
    let recentIncidents: Int
    let averageResponseTimeHours: Double
    let lastUpdated: String?
    
    enum CodingKeys: String, CodingKey {
        case totalIncidents = "total_incidents"
        case activeIncidents = "active_incidents"
        case resolvedIncidents = "resolved_incidents"
        case falseAlarms = "false_alarms"
        case incidentsByType = "incidents_by_type"
        case recentIncidents = "recent_incidents"
        case averageResponseTimeHours = "average_response_time_hours"
        case lastUpdated = "last_updated"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        totalIncidents = (try? container.decode(Int.self, forKey: .totalIncidents)) ?? 0
        activeIncidents = (try? container.decode(Int.self, forKey: .activeIncidents)) ?? 0
        resolvedIncidents = (try? container.decode(Int.self, forKey: .resolvedIncidents)) ?? 0
        falseAlarms = (try? container.decode(Int.self, forKey: .falseAlarms)) ?? 0
        incidentsByType = try? container.decode([String: Int].self, forKey: .incidentsByType)
        recentIncidents = (try? container.decode(Int.self, forKey: .recentIncidents)) ?? 0
        averageResponseTimeHours = (try? container.decode(Double.self, forKey: .averageResponseTimeHours)) ?? 0
        lastUpdated = try? container.decode(String.self, forKey: .lastUpdated)
    }
    
    func toDomain() -> DashboardAnalytics {
        DashboardAnalytics(
            totalIncidents: totalIncidents,
            activeIncidents: activeIncidents,
            resolvedIncidents: resolvedIncidents,
            totalRescueTeams: 0,
            totalAuthorities: 0,
            totalCitizens: 0,
            coastlineWatched: 0,
            activeAlerts: activeIncidents,
            pendingCount: activeIncidents,
            verifiedCount: 0,
            deployedCount: 0,
            resolvedCount: resolvedIncidents,
            avgVerificationTime: 0,
            avgResponseTime: averageResponseTimeHours,
            avgResolutionTime: 0,
            resolutionRate: totalIncidents > 0 ? Float(resolvedIncidents) / Float(totalIncidents) : 0
        )
    }
}

struct TimelineDataPointDTO: Codable {
    let date: String?
    let count: Int
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        date = try? container.decode(String.self, forKey: .date)
        count = (try? container.decode(Int.self, forKey: .count)) ?? 0
    }
    
    func toDomain() -> TimelineDataPoint {
        TimelineDataPoint(date: date ?? "", count: count)
    }
}

struct IncidentsTimelineDTO: Codable {
    let timeline: [TimelineDataPointDTO]?
    let periodDays: Int
    let startDate: String?
    let endDate: String?
    
    enum CodingKeys: String, CodingKey {
        case timeline
        case periodDays = "period_days"
        case startDate = "start_date"
        case endDate = "end_date"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        timeline = try? container.decode([TimelineDataPointDTO].self, forKey: .timeline)
        periodDays = (try? container.decode(Int.self, forKey: .periodDays)) ?? 30
        startDate = try? container.decode(String.self, forKey: .startDate)
        endDate = try? container.decode(String.self, forKey: .endDate)
    }
    
    func toDomain() -> IncidentsTimeline {
        let points = timeline?.map { $0.toDomain() } ?? []
        return IncidentsTimeline(
            dataPoints: points,
            totalCount: points.reduce(0) { $0 + $1.count },
            periodDays: periodDays
        )
    }
}

struct IncidentsDistributionDTO: Codable {
    let byType: [String: Int]?
    let byStatus: [String: Int]?
    let byUrgency: [String: Int]?
    
    enum CodingKeys: String, CodingKey {
        case byType = "by_type"
        case byStatus = "by_status"
        case byUrgency = "by_urgency"
    }
    
    func toDomain() -> IncidentsDistribution {
        IncidentsDistribution(
            byStatus: mapToDistributionItems(byStatus),
            byHazardType: mapToDistributionItems(byType),
            byUrgency: mapToDistributionItems(byUrgency)
        )
    }
    
    private func mapToDistributionItems(_ map: [String: Int]?) -> [DistributionItem] {
        guard let map = map else { return [] }
        let total = max(Float(map.values.reduce(0, +)), 1)
        return map.map { (label, value) in
            DistributionItem(
                label: label.replacingOccurrences(of: "_", with: " ").capitalized,
                value: value,
                percentage: Float(value) / total * 100
            )
        }
    }
}

struct GeographicIncidentDTO: Codable {
    let lat: Double
    let lng: Double
    let type: String?
    let status: String?
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        lat = (try? container.decode(Double.self, forKey: .lat)) ?? 0
        lng = (try? container.decode(Double.self, forKey: .lng)) ?? 0
        type = try? container.decode(String.self, forKey: .type)
        status = try? container.decode(String.self, forKey: .status)
    }
}

struct GeographicAnalyticsDTO: Codable {
    let incidents: [GeographicIncidentDTO]?
    let totalLocatedIncidents: Int
    
    enum CodingKeys: String, CodingKey {
        case incidents
        case totalLocatedIncidents = "total_located_incidents"
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        incidents = try? container.decode([GeographicIncidentDTO].self, forKey: .incidents)
        totalLocatedIncidents = (try? container.decode(Int.self, forKey: .totalLocatedIncidents)) ?? 0
    }
    
    func toDomain() -> GeographicAnalytics {
        let geoPoints = incidents?.map { incident in
            GeographicData(
                regionName: incident.type?.replacingOccurrences(of: "_", with: " ").capitalized ?? "Unknown",
                incidentCount: 1,
                latitude: incident.lat,
                longitude: incident.lng
            )
        } ?? []
        return GeographicAnalytics(
            regions: geoPoints,
            topRegions: Array(geoPoints.prefix(5))
        )
    }
}

struct UploadResponseDTO: Codable {
    let url: String
    let filename: String
}
